README.md
Dieses Projekt dient zum Monitoring und der Optimierung meiner BioWin2 von Windhager von 2018.

**Benutzung auf eigene Gefahr! Diese Skripte funktionieren für mich sind aber alles andere als gut getestet!**

Alle Features und Veränderungen welche ich in diesem Projekt vornehme nehmen keine Einfluss auf die BioWin2
bzw. wenn nur über zugängliche API's welche auch Dokumentiert sind.

# BioWin2

![BioWin2 Touch](screenshots/biowin2_touch.jpg)

## Zugangsdaten

Laut Anleitung, default:

* User: User ; Password: 123
* User: Service ; Password: 123

# Logging

Das wichtigste Thema dieses Projekts ist das Logging der Daten der Heizung um diese evtl. zu optimieren oder zumindest Probleme zu erkennen.
Mit unten genanntem Skript windhager-influx kann dies realisiert werden sofern eine oids.txt verfügbar ist.
Eine Liste mit allen im Setup verfügbaren Datenpunkten kann mit dem Skript windhager-getall.py erstellt werden:

    $ ./windhager-getall.py --windhager <ip-address> > oids_all.txt

Diese Liste sollte danach ausgedünnt werden da zumindest bei mir viele Datenpunkte nicht relevant oder zum Teil auch einfach doppelt vorhanden sind.
Beispielhaft ist die in meinem System verwendete oids.txt im Projekt hinterlegt.

## windhager-influx.py

Simples Skript welches die Datenpunkte welche in oids.txt (--oids) gelistet sind regelmäßig abfragt und in eine InfluxDB schreibt.
In meinem Fall werden die Daten per Grafana visualisiert:

Um dieses Skript der systemd automatisch zu starten kann die Datei `windhager-influx.service` genutzt werden. Dieses muss dazu nach z.B. `/etc/systemd/system/` kopiert werden.

![Grafana Beispiel](screenshots/20211129_grafana_beispiel.png)

# Warmwasser / Brennerstarts Optimierung

Skript: ww-override.py

## Problem

Bei normalem Winter-Betrieb Zündet die Heizung bei uns (ohne Pufferspeicher) 3-6x pro Tag für lediglich 1-3h.
Die längste Zeit ist morgens wenn sowohl FBH als auch WW hochheizen. Aber insbeondere hier ist folgendes Verhalten festzustellen:

* Sobald die WW Ladung abgeschlossen ist kommt es fast immer zum Ausbrand, da die Wärme nicht abgeführt werden kann und die Kesseltemperatur >80° wird.
* Kurz darauf (<1h) wird wieder gezündet

![Bild](screenshots/20211205_default_windhager.PNG)

Auch durch anpassen der Parameter für WW (Zeiten, Höhe usw.) konnte die Situation zwar "verschoben" werden aber es wurde auch immer deutlich der Komfort verschlechert. Auch mit dem Techniker wurde das Thema durchgegangen welcher auf einen Pufferspeicher verwies.

## Lösungsansatz (Winter d.h. FBH aktiv)

Dieses Skript nutzt Werte der BioWin um ein Relais (Tasmota) zu steuern welches parallel die WW-Ladepumpe schalten kann (die BioWin kann dies weiterhin auch noch!).
Die Schaltung hierfür sollte selbsterklärend sein.

Die BioWin WW-Temperaturangaben habe ich etwas verringert - aber durch die eingestellte Hysterese in der Heizung von -8K musste hier nur minimal angepasst werden.
Die WW-Anforderung der BioWin ist irgenwann vorbei. Normal würde hier der Überschwinger kommen. Hier schreitet nun dieses Skript ein in dem es folgendes tut:

Eingeschaltet wird die WW-Ladepumpe wenn:
* WW-Temperatur < ww_on oder
* Kesseltemperatur > kessel_max oder
* Kesselleitung < leistung_min

Abgeschaltet wird die WW-Ladepumpe wenn:
* Kessel Aus oder
* WW-Temperatur > ww_off oder
* Kesseltemperatur < kessel_min oder
* Kesselleitung > leistung_max

Dies hat zusätzlich zu der längeren Laufzeit den sehr scharmanten Vorteil, dass der WW-Speicher immer wieder etwas nachgeladen wird. Wird nun unter Tag viel WW Verbraucht muss die Heizung nicht erst zünden sondern schiebt einfach etwas Energie nach die vermutlich sowieso gerade zuviel ist.

## Ergebnisse:

### Erste manuelle Experimente

Hier wurde nur nebenzu die WW-Ladepumpe manuell geschaltet um ein erstes Gefühl zu bekommen. Aktiv war dies zwischen 8:00 bis 11:00 wobei ich ich den letzten Peak leider etwas zuspät abgefangen hatte. Am Nachmittag zwischen 15 und 17:00 Uhr. Auch hier habe ich beim letzten Peak zu langsam reagiert und gleich eine Zündung wegen veränderung des WW-Soll hinter her bekommen.

![Erste manuelle Versuche](screenshots/20211203_wwoverride_manul_first_try.png)

### Erste 1,5 Tage mit dem Skript

Zu sehen sind vier Starts in zwei Tagen. Der Morgentliche ist natürlich länger (6 Stunden, maximum der BioWin wie es scheint). Der am Nachmittag wird wohl nutzungsabhängig sein.
Noch zu untersuchen ist wie der Pelletverbrauch sich im Vergleich zu vorher entwickelt. Hierfür muss das ganze aber erst einmal eine Weile laufen.

![Erste zwei Tage](screenshots/20211211_ww_override.png)



# Bekannte Probleme

Hier werden bisher beobachtete Eigenheiten bzw. Probleme mit der Heizung aufgelistet und falls vorhanden Lösungen beschrieben.

## Zündung Ausbrand durch Brennstoffanforderung

System ist auf Brennstoffanforderung mit Vorgabezeit von 11:00 bis 15:00 Uhr eingestellt.
Zündet die Heizung kurz vor 15:00 (Ende der Freigabezeit) fällt ihr anscheinend kurz danach auf, dass sie noch Brennstoff braucht und geht direkt in den Ausbrand.

Antwort 13.12.2021 Windhager:

    wenn sie eine Freigabezeit von 12 – 15 Uhr hinterlegt habt wird der Kessel immer um kurz vor 15 Uhr in den Ausbrand gehen um eine Befüllung vor der Sperrzeit zu machen.

    Bild der Anleitung Kapitle "mit Freigabezeit" - markierung auf "Am Ende der Freigabezeit wird der Vorratsbehälter, falls notwendig, nochmals voll befüllt."

    Mit 3 Stunden Saugzeit kann es passieren das der Tagesbehälter leerläuft.

![Zündung Ausbrand](screenshots/20211212_ausbrand_bei_brennstoffanforderung.png)

## Viele OIDs nicht per Lookup Verfügbar

Es sind einige OIDs nicht im Rekursiven Lookup zugänglich.
Durch Brute-Force Zugriffe auf kleine Bereiche wurden bisher Etliche OID's gefunden welche allerdings bisher zum Großteil nicht zugeordnet sind. Eine Liste der bisher auf meiner Anlage entdeckten OIDs is in [oids_hidden.txt](oids_hidden.txt).

Bisher wurden folgende Werte zugeordnet:

* `1/60/0/23/88/0` : Saugzuggebläse-Soll
* `1/60/0/23/89/0` : Saugzuggebläse-Ist
* `1/60/0/43/42/0` : Füllstand-Tagesbehälter (%)
* `1/60/0/43/6/0` : Unterdruck-Soll (vermutlich)
* `1/60/0/43/7/0` : Unterdruck-Ist (vermutlich)
* `1/60/0/43/8/0` : Pelletförderung-pro-Stunde (vermutlich, kg als Einheit falsch)

Folgende Werte sind noch Vermutungen:

* `1/60/0/39/111/0` : könnte Ext. Wäremeanforderung Eingang (0..10V)
* `1/60/0/43/16/0` : kg seit Reinigung oder sowas? Sprang bei letzter Reinigung von ~450 auf 0, steigt seither
* `1/60/0/43/17/0` : kg bis Reiniung oder sowas? Sprang bei letzter Reinigung von 210 auf 650, sinkt seither eq. zu 43-16

## Undokummentierte Parameter im Webinterface

![Webinterface Unbekannte Parameter](screenshots/20211203_webif_unbekannt.png)

Im Web-Interface des InfoWin-Touch gibt es leider etliche Einträge "Unbekannte Ebene" bzw. "Unbekannter Wert" (siehe).
Leider konnte auch der Windhager-Support (technik@de.windhager.com) hier keine Abhilfe bieten da diesem hierzu selbst keine Dokumentation vorliegen würde. Es sei ein Feuerungsautomat verbaut welche in der Weboberfläche nicht eingepflegt wurde. Ganz wichtig: "Dei Funktion ist hier vollumfänglich gegeben" (das ist doch ein Kundenservice).

Aus diesem Grund wurden per windhager-influx.py für ein paar Wochen jedgliche Parameter aufgezeichnet und versucht deren Bedeutung heraus zu finden.

* Anzahl Werte gesamt: 144
* Anzahl Werte mit Namen: 111
* Anzahl Unbekannter Werte: 67

Auch sehr ärgerlich: Manche Parameter bei denen 'writeProtect' = False ist lassen sich nicht ändern. Stattdessen läuft der Request in einen Timeout mit Bad Gateway Error.



changelog.md
# Log der Einstellungsänderungen

## 2022 09 18
* change windhager-ww-override max WW to 57

## 2022 09 17
* 1/15/0/96/0 Betriebwahl 6 -> 1 (Heizprogramm 1)

## 2022 08 26
* Warmwasserprogramm 6:00 63° ; 10:00 40° (einmal aufheizen pro Tag)
* ww-override: Im Sommerbetrieb die Restwärme des Kessels in WW pumpen hinzugefügt

## 2021 12 11
* ww-override: Hinzugefügt und aktiviert um Brennerstarts zu minimieren
* Einstellungen insbesondere WW hierauf angepasst (Parameter nicht mehr klar)


Windhager.py
import sys
import logging
import json
import xml.etree.ElementTree as etree
import requests
from requests.auth import HTTPDigestAuth

class Windhager:
    """ Helper functions to interact with the Windhager InfoWin API
    as documented in self.ip/api-docs

    The default User/Password are given in the user manual.
    If using the myComfort App the password will be changed by Windhager!

    :param hostname: Host or IP of InfoWin device
    :param user: Username to Authenticate (from Usermanual)
    :param password: Password to Authenticate (from Usermanual)

    """

    LOOKUP_API = 'api/1.0/lookup'
    DATAPOINT_API = 'api/1.0/datapoint'
    DATAPOINTS_API = 'api/1.0/datapoints'
    OBJECT_API = 'api/1.0/object'

    def __init__(self, hostname, user = 'Service', password = '123', level='INFO'):
        self._host = hostname
        self._user = user
        self._password = password
        self._xml = None
        self.auth = HTTPDigestAuth(user, password)

        # configure logging
        logging.basicConfig()
        self.log = logging.getLogger(__name__)
        self.log.setLevel(level)

        # check connectivity to device
        self.get('api-docs/')

    def get(self, api, params = None, timeout=2):
        """ Issue a get request to InfoWin

        :param api: The API string to use
        :param params: Parameters to add to the GET request
        :param timeout: Timeout in seconds

        """

        r = requests.get(f"http://{self._host}/{api}", params = params, auth = self.auth)
        self.log.debug(f"GET http://{self._host}/{api} params={params} returned {r.status_code}")
        if r.status_code != 200:
            raise Exception
        try:
            return json.loads(r.text)
        except:
            return r.text

    def get_lookup(self, obj = None):
        """ Issue a request to lookup API

        From api-docs for GET:
            lookup/{subnetId}/{nodeId}/{fctNV}/0/{nvIndex}
        """

        if obj:
            return self.get(f"{self.LOOKUP_API}/{obj}")
        else:
            return self.get(self.LOOKUP_API)

    def get_lookup_all(self):
        """ Recursive lookup scan """
        ret = []

        # Get root node - should always be /1
        subnetIds = self.get_lookup()
        if not subnetIds:
            self.log.error("get_lookup_all: No subnetIds found!")
            return None

        # Iterate over all subnets
        for subnet in subnetIds:
            nodeIds = []
            nodeResp = self.get_lookup(subnet)
            for resp in nodeResp:
                if 'nodeId' in resp:
                    nodeIds.append(resp['nodeId'])

            # Iterate over all nodes
            for node in nodeIds:
                fctIds = []
                fctResp = self.get_lookup(f"{subnet}/{node}")
                if not 'functions' in fctResp:
                    self.log.error(f"Malformed fctResp, no functions found: {fctResp}")
                    return None
                for resp in fctResp['functions']:
                    if 'fctId' in resp:
                        fctIds.append(resp['fctId'])

                # Iterate over all functions
                for fct in fctIds:
                    subfctIds = []
                    try:
                        subfctResp = self.get_lookup(f"{subnet}/{node}/{fct}")
                    except:
                        continue
                    for resp in subfctResp:
                        if 'id' in resp:
                            subfctIds.append(resp['id'])

                    # Iterate over all subfunctions
                    for subfct in subfctIds:
                        try:
                            nvResp = self.get_lookup(f"{subnet}/{node}/{fct}/{subfct}")
                        except:
                            continue
                        for nv in nvResp:
                            ret.append(nv)

        return ret

    def get_object(self, obj, cache):
        """ Issue a request to object API

        From api-docs for GET:
            /object?OID=xx&cacheCtl={0,1}

        """

        return self.get(self.OBJECT_API, { 'OID': obj, 'cacheCtl': cache })

    def get_datapoint(self, obj):
        """ Issue a request to datapoint API

        From api-docs for GET:
            /datapoint/{subnetId}/{nodeId}/{fctId}/{groupId}/{memberId}/{varInst}
            /datapoint/{subnetId}/{nodeId}/{fctNV}/0/{nvIndex}/0

        """

        return self.get(f"{self.DATAPOINT_API}/{obj}")

    def get_datapoints(self):
        """ Issue a request to datapoints API

        TODO: per api-docs a param OID giving one or multiple OIDs should be supported.
            I did not get this working but always got Bad Gateway Errors. Need to revisit!

        """

        return self.get(f"{self.DATAPOINTS_API}")

    def get_datapoints_with_name(self):
        """ Get all (cached) datapoints and lookup the name -> text in the XML file """
        dps = get_datapoints()
        res = []
        for dp in dps:
            name = ""
            value = ""
            oid = ""
            if 'name' not in dp:
                res.append(dp)
                continue
            name = p['name']
            try:
                gN = int(p['name'].split('-')[0])
                mN = int(p['name'].split('-')[1])
            except:
                self.log.error(f"Malformed name: '{p['name']}' in datapoint {dp}")

            text = self.id_to_string(gN, mN)
            if text:
                p['text'] = text
            res.append(dp)
        return res

    @property
    def xml(self):
        """ Retrieve VarIdentTexte as parsed XML """
        if not self._xml:
            resp = self.get('res/xml/VarIdentTexte_de.xml')
            if not resp:
                self.log.error("Failed to get VarIdentTexte_de.xml")
            else:
                self._xml = etree.fromstring(resp.encode('utf-8'))
        return self._xml

    def id_to_string(self, gn, mn):
        """ Convert groupNr + memberNr to Text found in XML file if any """
        gn = int(gn)
        mn = int(mn)
        e = self.xml.find('gn[@id="' + str(gn) + '"]/mn[@id="' + str(mn) + '"]')
        if e is not None:
            if e.text is not None:
                return e.text.strip()
        return None

    def name_to_string(self, name):
        (gn, mn) = name.split('-')
        return self.id_to_string(gn, mn)


ww-override.py
#!/usr/bin/env python3

import time
import math
import os
import sys
import argparse
import requests
import Windhager

parser = argparse.ArgumentParser()
parser.add_argument('--windhager', type=str, required=True, help='Windhager IP/Host')
parser.add_argument('--tasmota', type=str, required=True, help='Tasmota IP/Host')
parser.add_argument('--wwmin', type=int, default=42, help='Minimum WW-Temperature')
parser.add_argument('--wwmax', type=int, default=70, help='Maximum WW-Temperature')
parser.add_argument('--kesselmax', type=int, default=75, help='Maximum Kessel-Temperature')
parser.add_argument('--kesselmin', type=int, default=65, help='Minimum Kessel-Temperature')
parser.add_argument('--leistungmin', type=int, default=45, help='Minimal Kessel-Leistung')
parser.add_argument('--leistungmax', type=int, default=95, help='Maximale Kessel-Leistung')
args = parser.parse_args()


def get_current_state(leistung, kessel, ww, now):
    # Kessel ist AN
    if leistung > 0 and kessel > 45:
        # Einschalten WW
        if ww < args.wwmin:
            return 'ON'
        if kessel > args.kesselmax and ww < args.wwmax:
            return 'ON'
        if leistung < args.leistungmin and ww < args.wwmax:
            return 'ON'

        # Ausschalten WW
        if kessel < args.kesselmin:
            return 'OFF'
        if leistung > args.leistungmax:
            return 'OFF'

    return 'OFF'

def set_tasmota_state(state):
    r = requests.get(f"http://{args.tasmota}/cm", params={'cmnd': f'Power1 {state}'})
    if r.status_code != 200:
        print(f"ERROR setting status to {state} on {args.tasmota}")

def init():
    Windhager.init(args.windhager)

def loop():
    print(f"\t\tOverride\tLeistung\tKessel\tWW")
    last_state = 'OFF'
    while True:
        leistung_ist = float(Windhager.get_datapoint('1/60/0/0/9/0')['value'])
        kessel_ist = float(Windhager.get_datapoint('1/60/0/0/7/0')['value'])
        ww_ist = float(Windhager.get_datapoint('1/15/0/0/4/0')['value'])
        now = time.localtime()
        nowstr = time.strftime("%H:%M:%S", now)

        new_state = get_current_state(leistung_ist, kessel_ist, ww_ist, now)
        set_tasmota_state(new_state)
        if new_state != last_state:
            last_state = new_state

        print(f"{nowstr}\t{new_state}\t\t{leistung_ist}\t\t{kessel_ist}\t{ww_ist}")

        time.sleep(15)
def main():
    mq = init()
    loop()

main()


windhager-getall.py
#!/usr/bin/env python3

import argparse
from Windhager import Windhager

parser = argparse.ArgumentParser()
parser.add_argument('--windhager', type=str, required=True, help='Windhager IP/Host')
parser.add_argument('--wuser', type=str, default='Service', help='Windhager Username')
parser.add_argument('--wpass', type=str, default='123', help='Windhager Password')
parser.add_argument('--debug', action='store_true', help='Activate Debug')
args = parser.parse_args()

def main():
    level = 'INFO'
    if args.debug:
        level = 'DEBUG'
    w = Windhager(args.windhager, user=args.wuser, password=args.wpass, level=level)

    dps = w.get_lookup_all()
    for dp in dps:
        if 'name' not in dp:
            continue
        if 'value' not in dp:
            continue
        key = f"{dp['groupNr']}-{dp['memberNr']}"
        try:
            value = float(dp['value'])
        except:
            value = None
        oid = dp['OID']
        name = w.id_to_string(dp['groupNr'], dp['memberNr'])
        if name:
            name = name.replace(' ','-')
        print(f"{oid.lstrip('/')},{key},{name},{value}")

if __name__ == '__main__':
    main()


windhager-influx.py
#!/usr/bin/env python3

import time
import argparse
from Windhager import Windhager
from influxdb import InfluxDBClient

parser = argparse.ArgumentParser()
parser.add_argument('--windhager', type=str, required=True, help='Windhager IP/Host')
parser.add_argument('--wuser', type=str, default='Service', help='Windhager Username')
parser.add_argument('--wpass', type=str, default='123', help='Windhager Password')
parser.add_argument('--influx', type=str, default='localhost', help='Influx Database Server')
parser.add_argument('--influxport', type=int, default=8086, help='Influx Database Server Port')
parser.add_argument('--db', type=str, required=True, help='Influx Database')
parser.add_argument('--dbuser', type=str, required=True, help='Influx Database User')
parser.add_argument('--dbpass', type=str, required=True, help='Influx Database Password')
parser.add_argument('--oids', type=str, default='oids.txt', help='File with list of OIDs')
parser.add_argument('--debug', action='store_true', help='Activate Debug')
args = parser.parse_args()

def push_influx(db, points):
    bodys = []
    for p in points:
        bodys.append({
            "measurement": f"windhager-{p[0]}",
            "fields": { 'value': p[1] }
        })
    db.write_points(bodys)

def loop(w, db, oids):
    while True:
        points = []
        for oid,entry in oids.items():
            if 'name' in entry:
                key = entry['name']
            else:
                continue
            d = w.get_datapoint(oid)
            if 'value' not in d:
                w.log.warning(f"Invalid data for {oid}: {dp}")
                continue

            try:
                value = float(d['value'])
            except:
                continue
            name = w.id_to_string(key.split('-')[0], key.split('-')[1])
            if name:
                name = name.replace(' ','-')
            else:
                if 'text' in entry:
                    name = entry['text']
                else:
                    name = "unknown"
            w.log.debug(f"got datapoint {name} {key} with {value}")
            points.append((f"{key}-{name}", value))

        push_influx(db, points)

        # Wait for next minute to begin
        time.sleep(5)
        while (int(time.time()) % 60):
            time.sleep(1)

def main():
    level = 'INFO'
    if args.debug:
        level = 'DEBUG'
    w = Windhager(args.windhager, user=args.wuser, password=args.wpass, level=level)
    db = InfluxDBClient(args.influx, args.influxport, database=args.db, username=args.dbuser, password=args.dbpass)
    oids = {}
    with open(args.oids, 'r') as fd:
        lines = fd.readlines()
        for line in lines:
            split = line.split(',')
            oids[split[0]] = {}
            if len(split) > 1:
                oids[split[0]]['name'] = split[1].rstrip()
            if len(split) > 2:
                oids[split[0]]['text'] = split[2].rstrip()

    w.log.info("Initialization done")

    loop(w, db, oids)

if __name__ == '__main__':
    main()



windhager-ww-override.py
#!/usr/bin/env python3

import argparse
import time
import requests
import paho.mqtt.client as paho
from Windhager import Windhager

parser = argparse.ArgumentParser()
parser.add_argument('--windhager', type=str, required=True, help='Windhager IP/Host')
parser.add_argument('--wuser', type=str, default='Service', help='Windhager Username')
parser.add_argument('--wpass', type=str, default='123', help='Windhager Password')
parser.add_argument('--mqtt', type=str, required=True, help='MQTT Broker')
parser.add_argument('--muser', type=str, required=True, help='MQTT Username')
parser.add_argument('--mpass', type=str, required=True, help='MQTT Password')
parser.add_argument('--mtopic', type=str, required=True, help='MQTT Topic')
parser.add_argument('--debug', action='store_true', help='Activate Debug')
args = parser.parse_args()

CONFIG = {
    'ww_on':        48,
    'ww_off':       54,
    'ww_max':       57,
    'kessel_min':   65,
    'kessel_max':   75,
    'leistung_min': 40,
    'leistung_max': 95
}

def get_current_state(w, last_state, leistung, kessel, ww, heizkreispumpe, now):
    #
    #  Pumpe ist AUS
    #
    if last_state == 'OFF':
        if leistung > 0 and kessel > ww and ww < CONFIG['ww_max']:
            if ww < CONFIG['ww_on']:
                w.log.info(f"WW < {CONFIG['ww_on']} -> turning ON")
                return 'ON'
            if kessel > CONFIG['kessel_max']:
                w.log.info(f"Kessel > {CONFIG['kessel_max']} -> turning ON")
                return 'ON'
            if leistung < CONFIG['leistung_min']:
                w.log.info(f"Leistung < {CONFIG['leistung_min']} -> turning ON")
                return 'ON'
        # Kessel aus und Heizkreispumpe aus -> Sommerbetrieb - Restwärme nutzen
        elif leistung == 0 and heizkreispumpe == 0 and kessel > ww + 5:
            w.log.info("Sommerbetrieb - Restwärme -> turning ON")
            return 'ON'
    #
    # WW Pumpe ist AN
    #
    elif last_state == 'ON':
        if leistung == 0 and (heizkreispumpe != 0 or kessel < ww + 3):
            w.log.info(f"Leistung == 0 -> turning OFF")
            return 'OFF'
        if ww > CONFIG['ww_max']:
            w.log.info(f"WW > {CONFIG['ww_max']} -> turning OFF")
            return 'OFF'
        if (ww > CONFIG['ww_off']) and (leistung > CONFIG['leistung_max']):
                w.log.info(f"Leistung > {CONFIG['leistung_max']} -> turning OFF")
                return 'OFF'
    return None

def set_mqtt_state(w, state):
    if state not in ['ON', 'OFF']:
        w.log.info(f"unknown state {state}")
        state = 'OFF'

    try:
        client = paho.Client()
        client.username_pw_set(username=args.muser, password=args.mpass)
        client.connect(args.mqtt, keepalive=10)
        client.publish(args.mtopic, state)
        client.disconnect()
    except Exception as e:
        w.log.error(f"setting status to {state} on {args.mtopic} failed!")
        w.log.error(f"{e}")

def loop(w):
    last_state = 'OFF'
    while True:
        leistung_ist = float(w.get_datapoint('1/60/0/0/9/0')['value'])
        kessel_ist = float(w.get_datapoint('1/60/0/0/7/0')['value'])
        ww_ist = float(w.get_datapoint('1/15/0/0/4/0')['value'])
        heizkreispumpe = float(w.get_datapoint('1/15/0/1/20/0')['value'])
        now = time.localtime()
        nowstr = time.strftime("%H:%M:%S", now)

        w.log.debug(f"last_state={last_state} leistung={leistung_ist} kessel={kessel_ist} ww={ww_ist} heizkreispumpe={heizkreispumpe}")

        new_state = get_current_state(w, last_state, leistung_ist, kessel_ist, ww_ist, heizkreispumpe, now)
        if new_state:
            w.log.info(f"set {args.mtopic} to {new_state}")
            if new_state != last_state:
                last_state = new_state
        # Just set it always to be sure not to get in unknown state for some time
        set_mqtt_state(w, last_state)
        time.sleep(20)

def main():
    level = 'INFO'
    if args.debug:
        level = 'DEBUG'
    w = Windhager(args.windhager, user=args.wuser, password=args.wpass, level=level)

    w.log.info("Initialization done")

    loop(w)

if __name__ == '__main__':
    main()



windhager-proxy.py
#!/usr/bin/env python3

import time
import json
import argparse
import logging
from Windhager import Windhager
from influxdb import InfluxDBClient
import paho.mqtt.client as paho

MQTT_BASE_TOPIC = "stat/windhager"
MQTT_STATE = "state"
MQTT_LWT = "lwt"

parser = argparse.ArgumentParser()
parser.add_argument('--windhager', type=str, required=True, help='Windhager IP/Host')
parser.add_argument('--wuser', type=str, default='Service', help='Windhager Username')
parser.add_argument('--wpass', type=str, default='123', help='Windhager Password')
parser.add_argument('--influx', type=str, default='localhost', help='Influx Database Server')
parser.add_argument('--influxport', type=int, default=8086, help='Influx Database Server Port')
parser.add_argument('--db', type=str, required=True, help='Influx Database')
parser.add_argument('--dbuser', type=str, required=True, help='Influx Database User')
parser.add_argument('--dbpass', type=str, required=True, help='Influx Database Password')
parser.add_argument('--mqtt', type=str, default='localhost', help='MQTT Broker URL')
parser.add_argument('--muser', type=str, required=True, help='MQTT User')
parser.add_argument('--mpass', type=str, required=True, help='MQTT Password')
parser.add_argument('--oids', type=str, default='oids.txt', help='File with list of OIDs')
parser.add_argument('--debug', action='store_true', help='Activate Debug')
args = parser.parse_args()

def poll_windhager_values(w, oids):
    points = []
    for oid,entry in oids.items():
        if 'name' not in entry:
            w.log.warning(f"Invalid config for {oid} : {entry}")
            continue
        key = entry['name']

        d = w.get_datapoint(oid)
        if 'value' not in d:
            w.log.warning(f"Invalid data for {oid}: {dp}")
            continue

        #TODO: float vs int vs string vs date?
        try:
            value = float(d['value'])
        except:
            continue
        #name = w.id_to_string(key.split('-')[0], key.split('-')[1])
        if 'text' in entry:
            name = entry['text'].replace(".", "").replace(" ", "_")
        else:
            name = "unknown"
        w.log.debug(f"got datapoint {name} {key} with {value}")
        points.append((key, name, value))

    return points

def influx_push(db, points):
    bodys = []
    for p in points:
        bodys.append({
            "measurement": f"windhager-{p[0]}-{p[1]}",
            "fields": { 'value': p[2] }
        })
    db.write_points(bodys)

def mqtt_push_values(mqtt, points):
    values = {}
    for p in points:
        text = p[1].lower().replace(".", "").replace("-", "_").replace("ö", "oe").replace("ä", "ae").replace("ü", "ue").replace("ß", "s")
        oid = p[0].lower().replace("-", "_")
        name =  f"{text}_{oid}"
        values[name] = p[2]

    mqtt.publish(f"{MQTT_BASE_TOPIC}/{MQTT_STATE}" , json.dumps(values))

def mqtt_push_discovery(mqtt, oids):
    for k, oidarr in oids.items():
        text = oidarr["text"].lower().replace(".", "").replace("-", "_").replace("ö", "oe").replace("ä", "ae").replace("ü", "ue").replace("ß", "s")
        oid = oidarr["name"].lower().replace("-", "_")
        name = f"windhager_{text}_{oid}"
        if "einheit" in oidarr:
            einheit = oidarr["einheit"]
        else:
            einheit = ""
        config = {
            "~": MQTT_BASE_TOPIC,
            "name": name,
            "suggested_area": "Windhager",
            "state_topic": f"~/{MQTT_STATE}",
            "value_template": "{{ value_json['" + f"{text}_{oid}" + "'] }}"
        }

        if einheit == "Binary":
            config["device_class"] = "power"
            config["payload_on"] = "1"
            config["payload_off"] = "0"
            mqtt.publish(f'homeassistant/binary_sensor/{name}/config', json.dumps(config))
        elif einheit == "State":
            if text == "betriebsphasen":
                config["value_template"] = "\
                    {% set mapper = { \
                        '0': 'Brenner gesperrt', \
                        '1': 'Selbsttest', \
                        '2': 'WE ausschalten', \
                        '3': 'Standby', \
                        '4': 'Brenner AUS', \
                        '5': 'Vorspülen', \
                        '6': 'Zündphase', \
                        '7': 'Flammstabilisierung', \
                        '8': 'Modulationsbetrieb', \
                        '9': 'Kessel gesperrt', \
                        '10': 'Standby Sperrzeit', \
                        '11': 'Gebläse AUS', \
                        '12': 'Verkleidungstür offen', \
                        '13': 'Zündung bereit', \
                        '14': 'Abbruch Zündphase', \
                        '15': 'Anheizvorgang', \
                        '16': 'Schichtlandung', \
                        '17': 'Ausbrand' \
                    } %} \
                    {% set state = int(value_json['" + f"{text}_{oid}" + "']) %} \
                    {{ mapper[state] if state in mapper else 'Unknown' }}"
            else:
                pass
            mqtt.publish(f'homeassistant/sensor/{name}/config', json.dumps(config))
        else:
            config["unit_of_measurement"] = einheit
            mqtt.publish(f'homeassistant/sensor/{name}/config', json.dumps(config))

        logging.debug(config)

def mqtt_on_publish(client, userdata, result):
    pass

def mqtt_on_message(client, userdata, message):
    pass

def mqtt_on_connect(client, userdata, flags, rc):
    logging.info("Connected to MQTT broker")
    client.publish(f"{MQTT_BASE_TOPIC}/{MQTT_LWT}", payload="Online", qos=0, retain=True)

def loop(w, db, mqtt, oids):
    last_discovery = 0
    while True:
        points = poll_windhager_values(w, oids)
        influx_push(db, points)

        if (time.time() - last_discovery > 900):
            logging.info("Pushing discovery data")
            mqtt_push_discovery(mqtt, oids)
            last_discovery = time.time()

        mqtt_push_values(mqtt, points)

        # Wait for next 1/4 minute
        time.sleep(1)
        while (int(time.time()) % 15):
            time.sleep(1)

def main():
    level = 'INFO'
    if args.debug:
        level = 'DEBUG'
    logging.basicConfig(level=level)
    w = Windhager(args.windhager, user=args.wuser, password=args.wpass, level=level)
    db = InfluxDBClient(args.influx, args.influxport, database=args.db, username=args.dbuser, password=args.dbpass)
    mqtt = paho.Client("windhager")
    mqtt.on_publish = mqtt_on_publish
    mqtt.on_message = mqtt_on_message
    mqtt.on_connect = mqtt_on_connect
    mqtt.username_pw_set(args.muser, args.mpass)
    mqtt.will_set(f"{MQTT_BASE_TOPIC}/{MQTT_LWT}", payload="offline", retain=True, qos=0)
    mqtt.connect(args.mqtt, keepalive=60)

    oids = {}
    with open(args.oids, 'r') as fd:
        lines = fd.readlines()
        for line in lines:
            split = line.split(',')
            oids[split[0]] = {}
            if len(split) > 1:
                oids[split[0]]['name'] = split[1].rstrip()
            if len(split) > 2:
                oids[split[0]]['text'] = split[2].rstrip()
            if len(split) > 4:
                oids[split[0]]['einheit'] = split[4].rstrip()

    logging.info("Initialization done")

    loop(w, db, mqtt, oids)

if __name__ == '__main__':
    main()

